pallets/asset/src/checkpoint.rs

```diff
@@ -0,0 +1,551 @@
+            // Compute the next timestamp.
+            let at = schedule
+                .next_checkpoint(now)
+                .ok_or(Error::<T>::FailedToComputeNextCheckpoint)?;
+
+            // Store removability.
+            ScheduleRemovable::insert((ticker, id), removable);
+
+            // Sort schedule into the queue.
+            let schedule = mk_schedule(at);
+            Schedules::insert(&ticker, {
+                let mut schedules = schedules;
+                add_schedule(&mut schedules, schedule);
+                schedules
+            });
+            schedule
+        };
+
+        Self::deposit_event(RawEvent::ScheduleCreated(did, ticker, schedule));
+        Ok(schedule)
+    }
+
+    /// The `actor` creates a checkpoint at `at` for `ticker`.
+    /// The ID of the new checkpoint is returned.
+    // TODO(Centril): privatize when dividend module is nixed.
+    pub fn create_at_by(
+        actor: IdentityId,
+        ticker: Ticker,
+        at: Moment,
+    ) -> Result<CheckpointId, DispatchError> {
+        let id = Self::next_ids(&ticker, 1)?;
+        Self::create_at(Some(actor), ticker, id, at);
+        Ok(id)
+    }
+
+    /// Creates a checkpoint at `at` for `ticker`, with the given, in advanced reserved, `id`.
+    /// The `actor` is the DID creating the checkpoint,
+    /// or `None` scheduling created the checkpoint.
+    ///
+    /// Creating a checkpoint entails:
+    /// - recording the total supply,
+    /// - mapping the the ID to the `time`.
+    fn create_at(actor: Option<IdentityId>, ticker: Ticker, id: CheckpointId, at: Moment) {
+        // Record total supply at checkpoint ID.
+        let supply = <Asset<T>>::token_details(ticker).total_supply;
+        <TotalSupply<T>>::insert(&(ticker, id), supply);
+
+        // Relate ID -> time.
+        Timestamps::insert(id, at);
+
+        // Emit event & we're done.
+        Self::deposit_event(RawEvent::CheckpointCreated(actor, ticker, id, supply, at));
+    }
+
+    /// Reserve `needed` amount of `CheckpointId`s and return the next ID to use.
+    fn next_ids(ticker: &Ticker, needed: u64) -> Result<CheckpointId, DispatchError> {
+        Total::try_mutate(ticker, |CheckpointId(slot)| {
+            slot.checked_add(needed)
+                .map(|new| CheckpointId(1.min(needed) + mem::replace(slot, new)))
+                .ok_or_else(|| Error::<T>::CheckpointOverflow.into())
+        })
+    }
+
+    /// Advance the checkpoint future ID counter and return the next ID to use.
+    fn next_schedule_id(ticker: &Ticker) -> Result<ScheduleId, DispatchError> {
+        ScheduleIds::try_mutate(ticker, |ScheduleId(id)| {
+            let next = id.checked_add(1).ok_or(Error::<T>::ScheduleOverflow)?;
+            Ok(ScheduleId(mem::replace(id, next)))
+        })
+    }
+
+    /// Returns the current UNIX time, i.e. seconds since UNIX epoch, 1970-01-01 00:00:00 UTC.
+    fn now_unix() -> Moment {
+        T::UnixTime::now().as_secs().saturated_into::<u64>()
```

Comment:
We use ms everywhere else, let's use ms here as well.

Reply:
Addressed through a separate story.